File: ai_server/services/fal_tryon_service.py
"""
fal.ai FASHN Virtual Try-On service.

Calls fal-ai/fashn/tryon/v1.5 via the official fal-client.
FAL_KEY must be set in the environment (via .env or shell).

Local file paths are uploaded to fal's storage using fal_client.upload_file()
so the external API can read them. Public URLs and base64 data URIs are
passed through directly.
"""

import os
import base64
import logging
import mimetypes
from typing import Optional

from schemas.tryon_schema import TryOnGenerateResponse

ALLOWED_CATEGORIES = {"tops", "bottoms", "one-pieces", "auto"}
ALLOWED_MODES = {"performance", "balanced", "quality"}
ALLOWED_GARMENT_PHOTO_TYPES = {"auto", "model", "flat-lay"}
ALLOWED_MODERATION_LEVELS = {"none", "permissive", "conservative"}
ALLOWED_OUTPUT_FORMATS = {"png", "jpeg"}
MAX_NUM_SAMPLES = 2

logger = logging.getLogger(__name__)


def get_default_fal_tryon_arguments() -> dict:
    return {
        "category": "auto",
        "mode": "quality",
        "garment_photo_type": "auto",
        "moderation_level": "permissive",
        "num_samples": 1,
        "segmentation_free": True,
        "output_format": "png",
    }


def prepare_fal_input(value: str) -> str:
    """
    Convert a local file path to a base64 data URI.
    Public https:// URLs and data: URIs are returned as-is.

    We intentionally avoid fal_client.upload_file() here because it requires fal
    storage/CDN permissions. Data URIs only require model inference access.
    """
    if value.startswith(("http://", "https://", "data:")):
        return value

    if not os.path.isfile(value):
        raise FileNotFoundError(f"Image file not found: {value}")

    mime_type, _ = mimetypes.guess_type(value)
    if not mime_type or not mime_type.startswith("image/"):
        mime_type = "image/jpeg"

    with open(value, "rb") as image_file:
        encoded = base64.b64encode(image_file.read()).decode("ascii")
    return f"data:{mime_type};base64,{encoded}"


def generate_tryon_with_fal(
    user_image_path: str,
    clothing_image_path: str,
    category: str = "auto",
    mode: str = "quality",
    garment_photo_type: str = "auto",
    moderation_level: str = "permissive",
    num_samples: int = 1,
    segmentation_free: bool = True,
    output_format: str = "png",
    seed: Optional[int] = None,
) -> TryOnGenerateResponse:
    fal_key = os.getenv("FAL_KEY", "").strip()
    if not fal_key:
        raise RuntimeError("FAL_KEY is not configured")

    os.environ["FAL_KEY"] = fal_key

    model = os.getenv("FAL_TRYON_MODEL", "fal-ai/fashn/tryon/v1.5")
    defaults = _get_configured_fal_tryon_arguments()
    overrides = _validate_fal_tryon_overrides(
        {
            "category": category,
            "mode": mode,
            "garment_photo_type": garment_photo_type,
            "moderation_level": moderation_level,
            "num_samples": num_samples,
            "segmentation_free": segmentation_free,
            "output_format": output_format,
            "seed": seed,
        }
    )
    final_args = {**defaults, **overrides}

    import fal_client

    model_url = prepare_fal_input(user_image_path)
    garment_url = prepare_fal_input(clothing_image_path)

    arguments = {
        "model_image": model_url,
        "garment_image": garment_url,
        **final_args,
    }

    logger.info(
        "Calling fal.ai try-on provider=fal model=%s arguments=%s",
        model,
        _sanitize_arguments_for_logging(arguments),
    )

    result = fal_client.subscribe(
        model,
        arguments=arguments,
        with_logs=True,
    )

    provider_error = _extract_provider_error(result)
    if provider_error:
        raise RuntimeError(provider_error)

    images = result.get("images") if isinstance(result, dict) else getattr(result, "images", None)
    if not images:
        raise RuntimeError("fal.ai try-on returned no images")

    first = images[0]
    image_url = first.get("url") if isinstance(first, dict) else getattr(first, "url", None)
    if not image_url:
        raise RuntimeError("fal.ai try-on image has no URL")

    return TryOnGenerateResponse(
        result_image_path=None,
        result_image_url=image_url,
        status="completed",
        provider="fal.ai",
        model=model,
        confidence_score=None,
        warnings=[],
        raw_images=images if isinstance(images[0], dict) else [{"url": i.url} for i in images],
    )


def _validated_choice(value: str, allowed: set[str], field: str) -> str:
    if not isinstance(value, str):
        raise ValueError(f"Invalid {field}. Allowed values: {', '.join(sorted(allowed))}")
    normalized = value.strip().lower()
    if normalized not in allowed:
        raise ValueError(f"Invalid {field}. Allowed values: {', '.join(sorted(allowed))}")
    return normalized


def _get_configured_fal_tryon_arguments() -> dict:
    defaults = get_default_fal_tryon_arguments()
    configured = {
        "category": os.getenv("FAL_TRYON_CATEGORY", defaults["category"]),
        "mode": os.getenv("FAL_TRYON_MODE", defaults["mode"]),
        "garment_photo_type": os.getenv("FAL_GARMENT_PHOTO_TYPE", defaults["garment_photo_type"]),
        "moderation_level": os.getenv("FAL_MODERATION_LEVEL", defaults["moderation_level"]),
        "num_samples": int(os.getenv("FAL_NUM_SAMPLES", str(defaults["num_samples"]))),
        "segmentation_free": _parse_bool(
            os.getenv("FAL_SEGMENTATION_FREE", str(defaults["segmentation_free"]))
        ),
        "output_format": os.getenv("FAL_OUTPUT_FORMAT", defaults["output_format"]),
    }
    return _validate_fal_tryon_overrides(configured)


def _validate_fal_tryon_overrides(values: dict) -> dict:
    validated = {}
    if "category" in values and values["category"] is not None:
        validated["category"] = _validated_choice(values["category"], ALLOWED_CATEGORIES, "category")
    if "mode" in values and values["mode"] is not None:
        validated["mode"] = _validated_choice(values["mode"], ALLOWED_MODES, "mode")
    if "garment_photo_type" in values and values["garment_photo_type"] is not None:
        validated["garment_photo_type"] = _validated_choice(
            values["garment_photo_type"],
            ALLOWED_GARMENT_PHOTO_TYPES,
            "garment_photo_type",
        )
    if "moderation_level" in values and values["moderation_level"] is not None:
        validated["moderation_level"] = _validated_choice(
            values["moderation_level"],
            ALLOWED_MODERATION_LEVELS,
            "moderation_level",
        )
    if "output_format" in values and values["output_format"] is not None:
        validated["output_format"] = _validated_choice(
            values["output_format"],
            ALLOWED_OUTPUT_FORMATS,
            "output_format",
        )
    if "num_samples" in values and values["num_samples"] is not None:
        validated["num_samples"] = _validated_num_samples(values["num_samples"])
    if "segmentation_free" in values and values["segmentation_free"] is not None:
        if not isinstance(values["segmentation_free"], bool):
            raise ValueError("segmentation_free must be a boolean")
        validated["segmentation_free"] = values["segmentation_free"]
    if "seed" in values and values["seed"] is not None:
        if not isinstance(values["seed"], int):
            raise ValueError("seed must be an integer")
        validated["seed"] = values["seed"]
    return validated


def _validated_num_samples(value: int) -> int:
    if not isinstance(value, int):
        raise ValueError("num_samples must be an integer")
    if value < 1 or value > MAX_NUM_SAMPLES:
        raise ValueError(f"num_samples must be between 1 and {MAX_NUM_SAMPLES}")
    return value


def _parse_bool(value: str) -> bool:
    normalized = value.strip().lower()
    if normalized in {"true", "1", "yes", "y"}:
        return True
    if normalized in {"false", "0", "no", "n"}:
        return False
    raise ValueError("FAL_SEGMENTATION_FREE must be true or false")


def _sanitize_arguments_for_logging(arguments: dict) -> dict:
    sanitized = dict(arguments)
    if "model_image" in sanitized:
        sanitized["model_image"] = "<image>"
    if "garment_image" in sanitized:
        sanitized["garment_image"] = "<image>"
    return sanitized


def _extract_provider_error(result) -> Optional[str]:
    if not isinstance(result, dict):
        return None
    error = result.get("error") or result.get("detail") or result.get("message")
    if not error:
        return None
    if isinstance(error, dict):
        return str(error.get("message") or error.get("detail") or error)
    return str(error)

File: FitApp/src/main/java/com/example/FitApp/ai/FalTryOnService.java
package com.example.FitApp.ai;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.http.client.SimpleClientHttpRequestFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;
import org.springframework.web.server.ResponseStatusException;
import tools.jackson.core.JacksonException;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

@Slf4j
@Service
public class FalTryOnService {

    private static final String FAL_QUEUE_BASE_URL = "https://queue.fal.run";
    private static final Set<String> ALLOWED_CATEGORIES = Set.of("tops", "bottoms", "one-pieces", "auto");
    private static final Set<String> ALLOWED_MODES = Set.of("performance", "balanced", "quality");
    private static final Set<String> ALLOWED_GARMENT_PHOTO_TYPES = Set.of("auto", "model", "flat-lay");
    private static final Set<String> ALLOWED_MODERATION_LEVELS = Set.of("none", "permissive", "conservative");
    private static final Set<String> ALLOWED_OUTPUT_FORMATS = Set.of("png", "jpeg");
    private static final int MAX_NUM_SAMPLES = 2;

    private final ObjectMapper objectMapper;
    private final RestClient falRestClient;

    public FalTryOnService(
            ObjectMapper objectMapper,
            @Value("${fal.http.connect-timeout-seconds:30}") long connectTimeoutSeconds,
            @Value("${fal.http.read-timeout-seconds:300}") long readTimeoutSeconds
    ) {
        this.objectMapper = objectMapper;

        SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
        requestFactory.setConnectTimeout(Duration.ofSeconds(connectTimeoutSeconds));
        requestFactory.setReadTimeout(Duration.ofSeconds(readTimeoutSeconds));
        this.falRestClient = RestClient.builder()
                .requestFactory(requestFactory)
                .build();
    }

    @Value("${fal.key:}")
    private String falKey;

    @Value("${fal.tryon.model:fal-ai/fashn/tryon/v1.5}")
    private String model;

    @Value("${fal.tryon.category:auto}")
    private String category;

    @Value("${fal.tryon.mode:quality}")
    private String mode;

    @Value("${fal.tryon.garment-photo-type:auto}")
    private String garmentPhotoType;

    @Value("${fal.tryon.moderation-level:permissive}")
    private String moderationLevel;

    @Value("${fal.tryon.num-samples:1}")
    private int numSamples;

    @Value("${fal.tryon.segmentation-free:true}")
    private boolean segmentationFree;

    @Value("${fal.tryon.output-format:png}")
    private String outputFormat;

    @Value("${fal.tryon.max-wait-seconds:360}")
    private long maxWaitSeconds;

    @Value("${fal.tryon.poll-interval-seconds:5}")
    private long pollIntervalSeconds;

    public boolean isConfigured() {
        return falKey != null && !falKey.isBlank();
    }

    public String getModel() {
        return model;
    }

    public FalTryOnResult generate(String userImagePath, String clothingImagePath) {
        if (!isConfigured()) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "FAL_KEY is not configured");
        }

        Map<String, Object> input = new HashMap<>();
        input.put("model_image", prepareFalInput(userImagePath));
        input.put("garment_image", prepareFalInput(clothingImagePath));
        input.putAll(defaultArguments());

        try {
            log.info("Calling fal.ai try-on model {} arguments {}", model, sanitizeForLogging(input));
            Map<String, Object> result = generateViaQueue(input);
            return toTryOnResult(result);
        } catch (ResponseStatusException e) {
            throw e;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "fal.ai try-on failed: interrupted");
        } catch (Exception e) {
            log.error("fal.ai try-on failed", e);
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "fal.ai try-on failed: " + rootMessage(e));
        }
    }

    private Map<String, Object> generateViaQueue(Map<String, Object> input) throws InterruptedException {
        String submitUrl = FAL_QUEUE_BASE_URL + "/" + model;
        Map<String, Object> submitted = executeFalJson(HttpMethod.POST, submitUrl, input);

        String requestId = stringValue(submitted.get("request_id"));
        String statusUrl = stringValue(submitted.get("status_url"));
        String responseUrl = stringValue(submitted.get("response_url"));
        if (requestId == null || statusUrl == null || responseUrl == null) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_GATEWAY,
                    "fal.ai queue submit response is missing request_id/status_url/response_url"
            );
        }

        log.info(
                "fal.ai try-on queued request {} position {} status_url={} response_url={}",
                requestId,
                submitted.get("queue_position"),
                statusUrl,
                responseUrl
        );

        waitForCompletion(requestId, statusUrl);
        return executeFalJson(HttpMethod.GET, responseUrl, null);
    }

    private void waitForCompletion(String requestId, String statusUrl) throws InterruptedException {
        Instant deadline = Instant.now().plus(Duration.ofSeconds(Math.max(1, maxWaitSeconds)));
        long pollMillis = Duration.ofSeconds(Math.max(1, pollIntervalSeconds)).toMillis();
        int emittedLogCount = 0;

        while (Instant.now().isBefore(deadline)) {
            String separator = statusUrl.contains("?") ? "&" : "?";
            Map<String, Object> statusResponse = executeFalJson(HttpMethod.GET, statusUrl + separator + "logs=1", null);

            emittedLogCount = logProviderMessages(statusResponse, emittedLogCount);
            String status = stringValue(statusResponse.get("status"));
            if ("COMPLETED".equals(status)) {
                return;
            }

            log.info("fal.ai try-on request {} status {}", requestId, status);
            Thread.sleep(pollMillis);
        }

        throw new ResponseStatusException(
                HttpStatus.GATEWAY_TIMEOUT,
                "fal.ai try-on timed out after " + maxWaitSeconds + " seconds"
        );
    }

    private int logProviderMessages(Map<String, Object> statusResponse, int alreadyEmitted) {
        List<?> logs = statusResponse.get("logs") instanceof List<?> list ? list : null;
        if (logs == null || logs.size() <= alreadyEmitted) {
            return alreadyEmitted;
        }

        for (int i = alreadyEmitted; i < logs.size(); i++) {
            log.info("fal.ai try-on provider log {}", logs.get(i));
        }
        return logs.size();
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> executeFalJson(HttpMethod method, String url, Map<String, Object> body) {
        String requestBody = body == null ? null : toJson(body);
        Instant startedAt = Instant.now();

        log.info("fal.ai request {} {}", method, url);
        log.info("fal.ai request headers {}", sanitizedRequestHeaders());
        if (requestBody != null) {
            log.info("fal.ai request body {}", sanitizeJsonForLogging(requestBody));
        }

        try {
            RestClient.RequestBodySpec requestSpec = falRestClient
                    .method(method)
                    .uri(url)
                    .headers(headers -> {
                        headers.set(HttpHeaders.AUTHORIZATION, "Key " + falKey.trim());
                        headers.setAccept(List.of(MediaType.APPLICATION_JSON));
                        if (requestBody != null) {
                            headers.setContentType(MediaType.APPLICATION_JSON);
                        }
                    });
            if (requestBody != null) {
                requestSpec.body(requestBody);
            }
            return requestSpec.exchange((request, response) -> parseFalResponse(startedAt, response));
        } catch (ResponseStatusException e) {
            throw e;
        } catch (Exception e) {
            long durationMs = Duration.between(startedAt, Instant.now()).toMillis();
            log.error("fal.ai request {} {} failed after {}ms", method, url, durationMs, e);
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "fal.ai request failed: " + rootMessage(e));
        }
    }

    private Map<String, Object> parseFalResponse(Instant startedAt, ClientHttpResponse response) throws IOException {
        String responseBody = readResponseBody(response);
        long durationMs = Duration.between(startedAt, Instant.now()).toMillis();
        log.info("fal.ai response {} {} after {}ms", response.getStatusCode().value(), response.getStatusText(), durationMs);
        log.info("fal.ai response headers {}", response.getHeaders());
        log.info("fal.ai response body {}", responseBody);

        if (response.getStatusCode().isError()) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_GATEWAY,
                    "fal.ai HTTP " + response.getStatusCode().value() + ": " + responseBody
            );
        }
        if (responseBody == null || responseBody.isBlank()) {
            return Collections.emptyMap();
        }
        return objectMapper.readValue(responseBody, new TypeReference<Map<String, Object>>() {});
    }

    private Map<String, String> sanitizedRequestHeaders() {
        return Map.of(
                HttpHeaders.AUTHORIZATION, "Key <redacted>",
                HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE,
                HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE
        );
    }

    private String readResponseBody(ClientHttpResponse response) throws IOException {
        byte[] bytes = response.getBody().readAllBytes();
        return new String(bytes, java.nio.charset.StandardCharsets.UTF_8);
    }

    private String toJson(Map<String, Object> body) {
        try {
            return objectMapper.writeValueAsString(body);
        } catch (JacksonException e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to serialize fal.ai request body");
        }
    }

    private String sanitizeJsonForLogging(String json) {
        return json
                .replaceAll("\"model_image\"\\s*:\\s*\"data:([^\"]{0,80})[^\"]*\"", "\"model_image\":\"data:$1...<redacted>\"")
                .replaceAll("\"garment_image\"\\s*:\\s*\"data:([^\"]{0,80})[^\"]*\"", "\"garment_image\":\"data:$1...<redacted>\"");
    }

    private String stringValue(Object value) {
        return value == null ? null : String.valueOf(value);
    }

    private Map<String, Object> defaultArguments() {
        Map<String, Object> arguments = new HashMap<>();
        arguments.put("category", validatedChoice(category, ALLOWED_CATEGORIES, "category"));
        arguments.put("mode", validatedChoice(mode, ALLOWED_MODES, "mode"));
        arguments.put("garment_photo_type", validatedChoice(
                garmentPhotoType,
                ALLOWED_GARMENT_PHOTO_TYPES,
                "garment_photo_type"
        ));
        arguments.put("moderation_level", validatedChoice(
                moderationLevel,
                ALLOWED_MODERATION_LEVELS,
                "moderation_level"
        ));
        arguments.put("num_samples", validatedNumSamples(numSamples));
        arguments.put("segmentation_free", segmentationFree);
        arguments.put("output_format", validatedChoice(outputFormat, ALLOWED_OUTPUT_FORMATS, "output_format"));
        return arguments;
    }

    private String prepareFalInput(String value) {
        if (value == null || value.isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Image path is required");
        }
        if (value.startsWith("http://") || value.startsWith("https://") || value.startsWith("data:")) {
            return value;
        }

        Path path = Path.of(value).toAbsolutePath().normalize();
        if (!Files.isRegularFile(path)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Image file not found: " + value);
        }

        try {
            String mimeType = Files.probeContentType(path);
            if (mimeType == null || !mimeType.startsWith("image/")) {
                mimeType = "image/jpeg";
            }
            String encoded = Base64.getEncoder().encodeToString(Files.readAllBytes(path));
            return "data:" + mimeType + ";base64," + encoded;
        } catch (IOException e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to read image file");
        }
    }

    private FalTryOnResult toTryOnResult(Map<String, Object> result) {
        String providerError = extractProviderError(result);
        if (providerError != null) {
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, providerError);
        }

        Object imagesValue = result == null ? null : result.get("images");
        if (!(imagesValue instanceof List<?> images) || images.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "fal.ai try-on returned no images");
        }

        String imageUrl = extractImageUrl(images.get(0));
        if (imageUrl == null || imageUrl.isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "fal.ai try-on image has no URL");
        }

        return new FalTryOnResult(
                imageUrl,
                "completed",
                "fal.ai",
                model,
                null,
                List.of(),
                new ArrayList<>(images)
        );
    }

    private String extractImageUrl(Object image) {
        if (image instanceof Map<?, ?> map && map.get("url") != null) {
            return String.valueOf(map.get("url"));
        }
        return null;
    }

    private String extractProviderError(Map<String, Object> result) {
        if (result == null) {
            return null;
        }
        Object error = firstNonNull(result.get("error"), result.get("detail"), result.get("message"));
        if (error instanceof Map<?, ?> map) {
            Object message = firstNonNull(map.get("message"), map.get("detail"));
            return message == null ? String.valueOf(error) : String.valueOf(message);
        }
        return error == null ? null : String.valueOf(error);
    }

    private Object firstNonNull(Object... values) {
        for (Object value : values) {
            if (value != null) {
                return value;
            }
        }
        return null;
    }

    private String validatedChoice(String value, Set<String> allowed, String field) {
        if (value == null) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Invalid " + field);
        }
        String normalized = value.trim().toLowerCase(Locale.ROOT);
        if (!allowed.contains(normalized)) {
            throw new ResponseStatusException(
                    HttpStatus.INTERNAL_SERVER_ERROR,
                    "Invalid " + field + ". Allowed values: " + String.join(", ", allowed)
            );
        }
        return normalized;
    }

    private int validatedNumSamples(int value) {
        if (value < 1 || value > MAX_NUM_SAMPLES) {
            throw new ResponseStatusException(
                    HttpStatus.INTERNAL_SERVER_ERROR,
                    "num_samples must be between 1 and " + MAX_NUM_SAMPLES
            );
        }
        return value;
    }

    private Map<String, Object> sanitizeForLogging(Map<String, Object> input) {
        Map<String, Object> sanitized = new HashMap<>(input);
        sanitized.put("model_image", "<image>");
        sanitized.put("garment_image", "<image>");
        return sanitized;
    }

    private String rootMessage(Throwable throwable) {
        Throwable current = throwable;
        while (current.getCause() != null) {
            current = current.getCause();
        }
        return current.getMessage() == null ? current.getClass().getSimpleName() : current.getMessage();
    }
}
